dev-docs/design/how-lit-html-works.md

```diff
@@ -0,0 +1,426 @@
+- `` html`<div>${html`hi`}</div>` ``, sets the TemplateResult value `` html`Hi` `` on the child part.
+- `` html`<div>${['h', 'i']}</div>` ``, sets the iterable on the child part.
+- `` html`<div>${document.createTextNode('hi')}</div>` ``, sets the Text node on the child part.
+
+At a very high level, `ChildPart` tracks the last value it was updated with via `_$committedValue`, skipping any work if the passed value is equal to the last committed value. This is how `ChildPart`s diff changes at the value level.
+If the `value` being set on the ChildPart is not the same as `_$committedValue`, then the value is processed (differently based on the type) into a Node or Nodes and is committed to the DOM. Then the `value` is set on `_$committedValue` so future calls to `_$setValue` can omit additional work if the value is unchanged.
+
+Primitive values set on a `ChildPart`, (`null`, `undefined`, `boolean`, `number`, `string`, `symbol`, or `bigint`), create or update a Text node. Committing a text node is done in `_commitText`, which is also where text sanitization occurrs.
+
+If the value is a `TemplateResult`, then ChildPart executes `_commitTemplateResult` which renders that `TemplateResult` into the `ChildPart`, recursively following the same steps of [preparing](#2i-prepare), [creating](#2ii-create) and [updating](#2iii-update) that `TemplateResult` described in this document.
+
+If the value is a Node, then the `ChildPart` clears any previously rendered nodes and inserts the Node value directly.
+
+#### AttributePart.\_$setValue
+
+In the single binding attribute value case such as `` html`<input value=${...}>` ``, an `AttributePart` also uses `_$committedValue` to diff against previously committed values and not do wasteful work. Values are committed by calling `this.element.setAttribute(name, value)`.
+
+`AttributePart` also handles the more complex cases with multiple bindings in attribute value position: `` html`<div class="${...} static-class ${...}"></div>` ``. Multiple values are also committed with a `setAttribute` call after the values are evaluated and joined together.
+
+#### PropertyPart.\_$setValue
+
+`PropertyPart` extends `AttributePart`, with the only difference being that committing the value to the live DOM is not via `setAttribute(name, value)`, but instead as a property via `this.element[name] = value`. E.g. `` html`<input .value=${'hi'}>` `` will commit the value `'hi'` with `inputEl.value = 'hi'`.
+
+#### BooleanAttributePart.\_$setValue
+
+`BooleanAttributePart` also extends `AttributePart` and overrides `_commitValue`. When a value is committed to a `BooleanAttributePart`, a falsy or `nothing` value will remove the attribute, and a truthy value will set the attribute with an empty string attribute value.
+
+#### EventPart.\_$setValue
+
+`EventPart` also extends `AttributePart`, and overrides `_$setValue` to take the user provided event listener and then call `this.element.addEventListener(attributeName, value)`.
```

Comment:
Is it worth mentioning that we bind `this`, and that we actually register a wrapper function so that updating the handler function doesn't need to call removeEventListener then addEventListener

Reply:
omg we do do the wrapper thing! TIL!

Reply:
Could you help me update the line: `The EventPart ensures that event listeners are added and cleaned up between renders and if the passed-in value changes.` to capture this information?

Reply:
How about something like this?

> The EventPart uses a listener object making it possible to update the listener without removing and adding a new one across re-renders even when the passed-in function changes. It will remove and add a new listener only if a listener object with different options compared to a previous render has been provided. A value of `null`, `undefined`, or `nothing` will remove any previously set listeners without adding a new listener.
>
> It also ensures that event listeners are called with the value of `this` set to a host object — often the host component rendering the template.

Also wondering if a link to https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#the_event_listener_callback might be useful as I don't know if an object with `handleEvent()` method serving as listener is common knowledge. Maybe that's TMI.

Reply:
+1 to all that @augustjk I think that's perfect

Reply:
Thank you!